### PR TITLE
[GEOT-6399] Arithmetic operations not supported with Complex Attributes

### PR DESCRIPTION
**Pull request by imranrajjad**
-added support for xpath numerical attributes for arithmetic operators
-integration tests added in app_schema

JIRA ticket : https://osgeo-org.atlassian.net/browse/GEOT-6399

### CODE REVIEW

File: modules/library/main/src/main/java/org/geotools/filter/Filters.java [hunk 1]
```diff
@@ -412,6 +414,17 @@ public static double number(Object value) {
                 throw new IllegalArgumentException("Unable to decode '" + text + "' as a number");
             }
         }
+        // if value is coming from ComplexAttribute through Xpath
+        if (value instanceof ComplexAttribute) {
```
Comment (aaime) [inline modules/library/main/src/main/java/org/geotools/filter/Filters.java]: Why does it even get there, the calling code should evaluate the expression/attribute before getting here.

Comment (imranrajjad) [reply modules/library/main/src/main/java/org/geotools/filter/Filters.java]: @aaime thanks for the review, so I am guessing it should be processed outside number method, where we can ensure whatever gets into this method will always be a number. Is that acceptable.?

**Comment by imranrajjad** [comment]
@aaime addressed requested changes, Expressions will be evaluated before Filters.number is called.

**Comment by aaime** [comment]
Looks better, but it's still ... odd. I had a look around and tried to figure out what was going on.
It seems that in 2011 support for handling complex values was added to the FilterAbstract and all its subclasses, via the eval(Expression, Object) method, which eventually calls onto unpack.

The move you did is not bad, but could be more complete and support the whole expression hierachy instead. 

Option 1:
- Copy eval(Expression, Object) to the Filters class, and delegate the existing method to it (instead 
- Add a delegating call to ExpressionAbstract too
- Use it in subclases to evaluate values, both in math expressions, but also, for example, in FunctionExpressionImpl.  Handling of classification functions is harder, I'd leave it to another ticket.

Option 2: deal with the unpacking madness at the source, in the relevant PropertyAccessor.
If I look at FeaturePropertyAccessor (which I believe it's the source of the issue, but it could as well be something else) I see it's completely ignoring the target class, and likely returning a ComplexAttribute when, say, a Number or a String was requested.
In this option, if the target class is set to anything other than Object, ComplexAttribute or null, I'd try to unpack the values and convert the to the requested target type.

@nmco what do you think?

**Comment by imranrajjad** [comment]
the subtle changes in second revision take inspiration from Option 1. By passing the intended class to evaluate method expression.evaluate(feature,<expected class>), we were able to obtain the numerical value nested inside Complex Feature attribute

**Comment by nmco** [comment]
So option 1 it is, what do you think from this approach @aaime ?

**Comment by imranrajjad** [comment]
@aaime @nmco 

added list supports, the expression will return a list in which each item will have passed through desired arithmetic operation.

if both expressions are lists, arithmetic operator will done on corresponding elements. For lists with un-equal size, the larger list will perform arithmetic operation as long as it can find a corresponding element, when the index has passed size of smaller list, unmodified element from larger list will be copied to result list.

File: modules/extension/app-schema/app-schema/src/test/java/org/geotools/appschema/filter/ArithmeticFilterTest.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        fSource = (FeatureSource<FeatureType, Feature>) dataAccess.getFeatureSource(FEATURE_TYPE);
+    }
+
+    @Test
+    public void testWithArithmeticOperator() throws IOException {
+
+        // evaluating first feature with Location accuracy value = 200
+
+        // 200 x 2 = 400
+        Filter arthimeticMultiplyFilter =
```
Comment (aaime) [inline modules/extension/app-schema/app-schema/src/test/java/org/geotools/appschema/filter/ArithmeticFilterTest.java]: arthimetic ;-) (see other variables too)

File: modules/library/main/src/main/java/org/geotools/filter/MathExpressionImpl.java [hunk 3]
```diff
@@ -109,4 +121,110 @@ protected Object number(double number) {
...
+                    .stream()
+                    .map(
+                            n ->
+                                    doArithmeticOperation(
+                                            n.doubleValue(), scalar.doubleValue(), operator))
+                    .collect(Collectors.toList());
+        }
+    }
+
+    private Double doArithmeticOperation(Double operand1, Double operand2, Operator operator) {
```
[the remark this answers is not in the document]
Comment (imranrajjad) [reply modules/library/main/src/main/java/org/geotools/filter/MathExpressionImpl.java]: understood, will delegate to subclass

File: modules/library/main/src/main/java/org/geotools/filter/MathExpressionImpl.java [hunk 4]
```diff
@@ -109,4 +121,110 @@ protected Object number(double number) {
...
+
+    protected Object getValue(org.opengis.filter.expression.Expression expression, Object feature) {
+        Object evaluation = expression.evaluate(feature);
+        return evaluation;
+    }
+
+    protected Object handleCollection(Object value1, Object value2, Operator operator) {
+        // case 1, both are collections
+        if (value1 instanceof Collection && value2 instanceof Collection) {
+            Collection<Number> result = new ArrayList<Number>();
```
[the remark this answers is not in the document]
Comment (imranrajjad) [reply modules/library/main/src/main/java/org/geotools/filter/MathExpressionImpl.java]: understood, will remove

File: modules/extension/app-schema/app-schema/src/test/java/org/geotools/appschema/filter/ArithmeticFilterTest.java (hunk 2 above)
Comment (imranrajjad) [reply modules/extension/app-schema/app-schema/src/test/java/org/geotools/appschema/filter/ArithmeticFilterTest.java]: will fix spelling typos

**Comment by imranrajjad** [comment]
@aaime addressed requested changes, hopefully this should be it :)

File: modules/extension/app-schema/app-schema/src/test/java/org/geotools/appschema/filter/ArithmeticFilterTest.java (hunk 2 above)
Comment (aaime) [reply modules/extension/app-schema/app-schema/src/test/java/org/geotools/appschema/filter/ArithmeticFilterTest.java]: Spelling errors still seem to be here?

Comment (imranrajjad) [reply modules/extension/app-schema/app-schema/src/test/java/org/geotools/appschema/filter/ArithmeticFilterTest.java]: missed those, will fix

**Comment by imranrajjad** [comment]
22.x backport PR https://github.com/geotools/geotools/pull/2616

**Pull request merged**